### PR TITLE
feat : add storage diff rpc

### PR DESCRIPTION
**Pull request by dvcpull**
RPC to return changes in storage between two blocks,

Example : 
`curl -sS -H "Content-Type: application/json" -d '{"id":1, "jsonrpc":"2.0", "method": "state_getStorageDiff", "params": ["0xaa254e5ff33843f4abfb2527c7273239f395f7f7da10b6018646f4ae7257f913", "0xb2990969fd7771959f534d3d985603b5632fb6a3faf628d7ba79c564424de7ae"]}' http://localhost:9944/`

Result : 
```
{"jsonrpc":"2.0","result":[["0x26aa394eea5630e07c48ae0c9558cef702a5c1b19ab7a04f536c519aca4983ac","0x33000000"],["0x26aa394eea5630e07c48ae0c9558cef78a42f33323cb5ced3b44dd825fda9fcc","0xda6e43000582e224d717d60d1fa65ef7bc5b05ccb4ff412c67f8d334a86c8cbd"],["0x26aa394eea5630e07c48ae0c9558cef799e7f93fc6a98f0874fd057f111c4d2d","0x04066175726120baf1d91000000000"],["0x26aa394eea5630e07c48ae0c9558cef7a44704b568d21667356a5a050c1187468e233f9722d30ebf31000000","0xaa254e5ff33843f4abfb2527c7273239f395f7f7da10b6018646f4ae7257f913"],["0x26aa394eea5630e07c48ae0c9558cef7a44704b568d21667356a5a050c118746926f3965faf45f0f16000000","0x403edf6b6b63e8f0402b4815ef63efd29628e29e92c9931012d7470b7730cd0c"],["0x26aa394eea5630e07c48ae0c9558cef7a44704b568d21667356a5a050c1187469d7a9a93920005ae1d000000","0xe7cd5c4f22a46446a36066dffb9c303ceb8d00c2d0a421c56d6ee1b5daf86593"],["0x26aa394eea5630e07c48ae0c9558cef7a44704b568d21667356a5a050c1187469eb2dcce60f37a2702000000","0x75d57415c5640f1c184a9424e01b4979a74f0b6f19a6e41bee3df52eb3b8d4a3"],["0x26aa394eea5630e07c48ae0c9558cef7a44704b568d21667356a5a050c118746a6b274250e6753f00a000000","0xfdd294deb2e99978a9a24eb8eee81de72d839f6c27588ba255e29050e9799cff"],....
```

**Performance when tested :**

**Env** : Substrate node template runtime
**Tested using** : `curl -sS -s -w 'Establish Connection: %{time_connect}s\nTTFB: %{time_starttransfer}s\nTotal: %{time_total}s\n' -H "Content-Ty....`
**Result** :
```
Establish Connection: 0.006368s
TTFB: 0.024987s
Total: 0.040980s
```

### CODE REVIEW

File: substrate/client/rpc/src/state/state_full.rs [hunk 1]
```diff
@@ -217,6 +216,35 @@ where
...
+				storage_diff.push((key, new_storage))
+			} else {
+				let start_storage_val = self.client.storage(start, &key).map_err(client_err)?;
+				let end_storage_val = self.client.storage(end, &key).map_err(client_err)?;
+
+				if start_storage_val != end_storage_val {
+					storage_diff.push((key, end_storage_val))
+				}
+			}
+		}
```
Comment (mezrin) [inline substrate/client/rpc/src/state/state_full.rs:242]: This algo ignores keys that existed in `start` and then were deleted.
Such keys must be added to the result: `storage_diff.push((deleted_key, Null))`

**Review by mezrin — CHANGES_REQUESTED** [review]

File: substrate/client/rpc/src/state/state_full.rs (hunk 1 above)
Comment (dvcpull) [reply substrate/client/rpc/src/state/state_full.rs:242]: In this case, 
```
start_storage_value = Some(xxx)
end_storage_value = None
start_storage_value != end_storage_value (so will be pushed to storage_diff)
```

client.storage returns Option (https://github.com/traittech/polkadot-sdk/blob/122086d3d582b1e86190507f3a377667e66c729c/substrate/client/rpc/src/state/state_full.rs#L250C1-L250C1)

Comment (mezrin) [reply substrate/client/rpc/src/state/state_full.rs:242]: no, it won't
You iterate the `end_keys`: `for key in end_keys {`.
If a key-value pair was deleted - it won't be included in `end_keys`

Comment (mezrin) [reply substrate/client/rpc/src/state/state_full.rs:242]: Screenshot 2023-10-12 at 12 55 27

marked with yellow the missed keys

Comment (dvcpull) [reply substrate/client/rpc/src/state/state_full.rs:242]: Ah I see your point

**Pull request merged**